File: SM_MentalHealthApp.Client/Services/AgoraService.cs
using Microsoft.JSInterop;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Net.Http.Json;

namespace SM_MentalHealthApp.Client.Services
{
    public interface IAgoraService
    {
        Task<bool> InitializeAsync(string appId);
        Task<bool> JoinChannelAsync(string channelName, string token, uint uid);
        Task<bool> JoinChannelAsync(string channelName, string token, string appId, uint uid, bool isVideoCall);
        Task LeaveChannelAsync();
        Task EnableLocalVideoAsync(bool enable);
        Task EnableLocalAudioAsync(bool enable);
        Task SwitchCameraAsync();
        Task MuteLocalAudioAsync(bool mute);
        Task MuteLocalVideoAsync(bool mute);
        Task DestroyAsync();
        bool IsInitialized { get; }
        bool IsInCall { get; }
        string CurrentChannel { get; }
        uint CurrentUid { get; }
        event Action<uint> OnUserJoined;
        event Action<uint> OnUserLeft;
        event Action<string> OnConnectionStateChanged;
        event Action<string> OnError;
        //Task<string> GetAgoraTokenAsync(string channelName);
        Task<(string Token, string AppId)> GetAgoraTokenAsync(string channelName, uint uid = 0);
    }

    public class AgoraService : IAgoraService, IAsyncDisposable
    {
        private readonly IJSRuntime _jsRuntime;
        private readonly HttpClient _httpClient;
        private readonly IAuthService _authService;
        private bool _isInitialized = false;
        private bool _isInCall = false;
        private string _currentChannel = string.Empty;
        private uint _currentUid = 0;

        public AgoraService(IJSRuntime jsRuntime, HttpClient httpClient, IAuthService authService)
        {
            _jsRuntime = jsRuntime;
            _httpClient = httpClient;
            _authService = authService;
        }

        public bool IsInitialized => _isInitialized;
        public bool IsInCall => _isInCall;
        public string CurrentChannel => _currentChannel;
        public uint CurrentUid => _currentUid;

        public event Action<uint>? OnUserJoined;
        public event Action<uint>? OnUserLeft;
        public event Action<string>? OnConnectionStateChanged;
        public event Action<string>? OnError;

        public async Task<bool> InitializeAsync(string appId)
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraInitialize", appId, DotNetObjectReference.Create(this));
                _isInitialized = true;
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to initialize Agora: {ex.Message}");
                return false;
            }
        }

        public async Task<bool> JoinChannelAsync(string channelName, string token, uint uid)
        {
            // This overload is deprecated - should use the one with appId
            // For backward compatibility, try to get appId from token response
            var (newToken, appId) = await GetAgoraTokenAsync(channelName, uid);
            if (string.IsNullOrEmpty(newToken))
            {
                // Fallback to default App ID
                appId = "efa11b3a7d05409ca979fb25a5b489ae";
            }
            return await JoinChannelAsync(channelName, string.IsNullOrEmpty(newToken) ? token : newToken, appId, uid, true); // Default to video call
        }

        public async Task<bool> JoinChannelAsync(string channelName, string token, string appId, uint uid, bool isVideoCall)
        {
            try
            {
                if (!_isInitialized)
                {
                    throw new InvalidOperationException("Agora not initialized");
                }

                Console.WriteLine($"🎯 Joining Agora channel: {channelName} with UID: {uid}, Video: {isVideoCall}, AppId: {appId}");
                await _jsRuntime.InvokeVoidAsync("agoraJoinChannel", appId, channelName, token, uid, isVideoCall);
                _isInCall = true;
                _currentChannel = channelName;
                _currentUid = uid;
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to join channel........Iqbal2: {ex.Message}");
                return false;
            }
        }

        public async Task LeaveChannelAsync()
        {
            try
            {
                if (_isInCall)
                {
                    await _jsRuntime.InvokeVoidAsync("agoraLeaveChannel");
                    _isInCall = false;
                    _currentChannel = string.Empty;
                    _currentUid = 0;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to leave channel: {ex.Message}");
            }
        }

        public async Task EnableLocalVideoAsync(bool enable)
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraEnableLocalVideo", enable);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to toggle local video: {ex.Message}");
            }
        }

        public async Task EnableLocalAudioAsync(bool enable)
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraEnableLocalAudio", enable);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to toggle local audio: {ex.Message}");
            }
        }

        public async Task SwitchCameraAsync()
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraSwitchCamera");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to switch camera: {ex.Message}");
            }
        }

        public async Task MuteLocalAudioAsync(bool mute)
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraMuteLocalAudio", mute);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to mute/unmute audio: {ex.Message}");
            }
        }

        public async Task MuteLocalVideoAsync(bool mute)
        {
            try
            {
                await _jsRuntime.InvokeVoidAsync("agoraMuteLocalVideo", mute);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to mute/unmute video: {ex.Message}");
            }
        }

        public async Task DestroyAsync()
        {
            try
            {
                if (_isInCall)
                {
                    await LeaveChannelAsync();
                }
                await _jsRuntime.InvokeVoidAsync("agoraDestroy");
                _isInitialized = false;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to destroy Agora: {ex.Message}");
            }
        }

        // public async Task<string> GetAgoraTokenAsync(string channelName)
        // {
        //     try
        //     {
        //         Console.WriteLine($"🎯 Getting Agora token for channel: {channelName}");

        //         // Get token from storage directly
        //         var token = await _jsRuntime.InvokeAsync<string?>("getToken");
        //         if (string.IsNullOrEmpty(token))
        //         {
        //             Console.WriteLine("❌ User not authenticated - no token in storage");
        //             throw new Exception("User not authenticated");
        //         }

        //         Console.WriteLine($"✅ Auth token found: {token.Substring(0, Math.Min(20, token.Length))}...");

        //         var request = new { channelName = channelName, expirationTimeInSeconds = 3600 };

        //         // Set authorization header
        //         _httpClient.DefaultRequestHeaders.Authorization =
        //             new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);

        //         Console.WriteLine($"🌐 Calling API: {_httpClient.BaseAddress}api/agora/token");
        //         var response = await _httpClient.PostAsJsonAsync("/api/agora/token", request);

        //         Console.WriteLine($"📡 API Response Status: {response.StatusCode}");

        //         if (!response.IsSuccessStatusCode)
        //         {
        //             var errorContent = await response.Content.ReadAsStringAsync();
        //             Console.WriteLine($"❌ API Error: {errorContent}");
        //             throw new Exception($"API call failed: {response.StatusCode} - {errorContent}");
        //         }

        //         var responseContent = await response.Content.ReadAsStringAsync();
        //         Console.WriteLine($"📄 API Response: {responseContent}");

        //         var result = await response.Content.ReadFromJsonAsync<AgoraTokenResponse>();
        //         Console.WriteLine($"✅ Parsed token: {(result?.Token != null ? result.Token.Substring(0, Math.Min(20, result.Token.Length)) : "null")}...");

        //         return result?.Token ?? string.Empty;
        //     }
        //     catch (Exception ex)
        //     {
        //         Console.WriteLine($"❌ Failed to get Agora token: {ex.Message}");
        //         Console.WriteLine($"❌ Stack trace: {ex.StackTrace}");
        //         return string.Empty;
        //     }
        // }

        public async Task<(string Token, string AppId)> GetAgoraTokenAsync(string channelName, uint uid = 0)
        {
            try
            {
                // ✅ Validate channel name before making request
                if (string.IsNullOrWhiteSpace(channelName))
                {
                    Console.WriteLine("❌ Invalid channel name: channel name is empty or null");
                    throw new ArgumentException("Channel name cannot be empty or null", nameof(channelName));
                }

                Console.WriteLine($"🎯 Getting Agora token for channel: {channelName}");

                // ✅ 1. Get user JWT token from browser storage
                var authToken = await _jsRuntime.InvokeAsync<string?>("getToken");
                if (string.IsNullOrEmpty(authToken))
                {
                    Console.WriteLine("❌ User not authenticated - no token in storage");
                    throw new Exception("User not authenticated");
                }

                Console.WriteLine($"✅ Auth token found: {authToken.Substring(0, Math.Min(20, authToken.Length))}...");

                // ✅ 2. Use .NET Server for token generation (same as main API)
                var url = $"api/realtime/token?channel={Uri.EscapeDataString(channelName)}&uid={uid}";
                Console.WriteLine($"🌐 Calling .NET Server: {_httpClient.BaseAddress}{url}");

                // ✅ 3. Attach Authorization header for .NET server
                _httpClient.DefaultRequestHeaders.Authorization =
                    new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", authToken);

                // ✅ 4. Make the GET request and parse JSON directly
                var result = await _httpClient.GetFromJsonAsync<AgoraTokenResponse>(url);

                // ✅ 5. Validate and return token with App ID
                if (result == null || string.IsNullOrEmpty(result.Token))
                {
                    Console.WriteLine("❌ Invalid or empty Agora token received from token server");
                    return (string.Empty, string.Empty);
                }

                var appId = result.AgoraAppId;
                if (string.IsNullOrEmpty(appId))
                {
                    // Fallback to default App ID if not in response
                    appId = "efa11b3a7d05409ca979fb25a5b489ae";
                    Console.WriteLine("⚠️ App ID not in response, using default");
                }

                Console.WriteLine($"✅ Agora token received from .NET server (cached={result.Cached}, appId={appId}): {result.Token.Substring(0, Math.Min(20, result.Token.Length))}...");
                return (result.Token, appId);
            }
            catch (HttpRequestException httpEx)
            {
                Console.WriteLine($"❌ Network error fetching Agora token: {httpEx.Message}");
                return (string.Empty, string.Empty);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Failed to get Agora token: {ex.Message}");
                Console.WriteLine($"❌ Stack trace: {ex.StackTrace}");
                return (string.Empty, string.Empty);
            }
        }


        // JavaScript interop callbacks
        [JSInvokable]
        public void OnUserJoinedCallback(uint uid)
        {
            OnUserJoined?.Invoke(uid);
        }

        [JSInvokable]
        public void OnUserLeftCallback(uint uid)
        {
            OnUserLeft?.Invoke(uid);
        }

        [JSInvokable]
        public void OnConnectionStateChangedCallback(string state)
        {
            OnConnectionStateChanged?.Invoke(state);
        }

        [JSInvokable]
        public void OnErrorCallback(string error)
        {
            OnError?.Invoke(error);
        }

        public async ValueTask DisposeAsync()
        {
            await DestroyAsync();
        }
    }

    public class AgoraTokenResponse
    {
        [JsonPropertyName("agoraAppId")]
        public string AgoraAppId { get; set; } = string.Empty;

        [JsonPropertyName("token")]
        public string Token { get; set; } = string.Empty;

        [JsonPropertyName("cached")]
        public bool Cached { get; set; }

        [JsonPropertyName("channelName")]
        public string ChannelName { get; set; } = string.Empty;

        [JsonPropertyName("uid")]
        public uint Uid { get; set; }

        [JsonPropertyName("expirationTime")]
        public DateTime? ExpirationTime { get; set; }
    }
}

File: MentalHealthMobile/MentalHealthMobileSimple/src/services/AgoraService.js
import {
  RtcEngine,
  RtcLocalView,
  RtcRemoteView,
  VideoRenderMode,
  ClientRole,
  ChannelProfile,
} from 'react-native-agora';

class AgoraService {
  constructor() {
    this.engine = null;
    this.isInitialized = false;
    this.isInCall = false;
    this.currentChannel = null;
    this.currentUid = null;
    this.listeners = {
      onUserJoined: null,
      onUserLeft: null,
      onError: null,
    };
  }

  async initialize(appId) {
    try {
      console.log('🚀 Mobile: Initializing Agora with App ID:', appId);
      
      // Create RTC engine
      this.engine = await RtcEngine.create(appId);
      
      // Enable video
      await this.engine.enableVideo();
      
      // Set channel profile to communication
      await this.engine.setChannelProfile(ChannelProfile.Communication);
      
      // Set client role to broadcaster (for two-way communication)
      await this.engine.setClientRole(ClientRole.Broadcaster);
      
      // Set up event listeners
      this.engine.addListener('UserJoined', (uid, elapsed) => {
        console.log('📱 Mobile: User joined:', uid);
        if (this.listeners.onUserJoined) {
          this.listeners.onUserJoined(uid);
        }
      });
      
      this.engine.addListener('UserOffline', (uid, reason) => {
        console.log('📱 Mobile: User left:', uid, reason);
        if (this.listeners.onUserLeft) {
          this.listeners.onUserLeft(uid);
        }
      });
      
      this.engine.addListener('Error', (err) => {
        console.error('❌ Mobile: Agora error:', err);
        if (this.listeners.onError) {
          this.listeners.onError(err);
        }
      });
      
      this.engine.addListener('JoinChannelSuccess', (channel, uid, elapsed) => {
        console.log('✅ Mobile: Joined channel successfully:', channel, uid);
        this.isInCall = true;
        this.currentChannel = channel;
        this.currentUid = uid;
      });
      
      this.isInitialized = true;
      console.log('✅ Mobile: Agora initialized successfully');
      return true;
    } catch (error) {
      console.error('❌ Mobile: Failed to initialize Agora:', error);
      return false;
    }
  }

  async joinChannel(channelName, token, uid = 0) {
    try {
      if (!this.isInitialized || !this.engine) {
        throw new Error('Agora not initialized');
      }

      console.log('🎯 Mobile: Joining channel:', channelName, 'with UID:', uid);
      
      // Join channel
      await this.engine.joinChannel(token, channelName, null, uid);
      
      return true;
    } catch (error) {
      console.error('❌ Mobile: Failed to join channel:....other1', error);
      return false;
    }
  }

  async leaveChannel() {
    try {
      if (this.engine && this.isInCall) {
        console.log('👋 Mobile: Leaving channel');
        await this.engine.leaveChannel();
        this.isInCall = false;
        this.currentChannel = null;
        this.currentUid = null;
      }
    } catch (error) {
      console.error('❌ Mobile: Failed to leave channel:', error);
    }
  }

  async enableLocalVideo(enable) {
    try {
      if (this.engine) {
        if (enable) {
          await this.engine.enableLocalVideo(true);
        } else {
          await this.engine.enableLocalVideo(false);
        }
        return true;
      }
      return false;
    } catch (error) {
      console.error('❌ Mobile: Failed to toggle video:', error);
      return false;
    }
  }

  async enableLocalAudio(enable) {
    try {
      if (this.engine) {
        if (enable) {
          await this.engine.enableLocalAudio(true);
        } else {
          await this.engine.enableLocalAudio(false);
        }
        return true;
      }
      return false;
    } catch (error) {
      console.error('❌ Mobile: Failed to toggle audio:', error);
      return false;
    }
  }

  async muteLocalAudio(mute) {
    try {
      if (this.engine) {
        await this.engine.muteLocalAudioStream(mute);
        return true;
      }
      return false;
    } catch (error) {
      console.error('❌ Mobile: Failed to mute audio:', error);
      return false;
    }
  }

  async muteLocalVideo(mute) {
    try {
      if (this.engine) {
        await this.engine.muteLocalVideoStream(mute);
        return true;
      }
      return false;
    } catch (error) {
      console.error('❌ Mobile: Failed to mute video:', error);
      return false;
    }
  }

  async switchCamera() {
    try {
      if (this.engine) {
        await this.engine.switchCamera();
        return true;
      }
      return false;
    } catch (error) {
      console.error('❌ Mobile: Failed to switch camera:', error);
      return false;
    }
  }

  setEventListeners(listeners) {
    this.listeners = { ...this.listeners, ...listeners };
  }

  async destroy() {
    try {
      if (this.isInCall) {
        await this.leaveChannel();
      }
      if (this.engine) {
        await RtcEngine.destroy();
        this.engine = null;
        this.isInitialized = false;
      }
    } catch (error) {
      console.error('❌ Mobile: Failed to destroy Agora:', error);
    }
  }

  // Helper method to get Agora token from server
  async getAgoraToken(channelName, authToken) {
    try {
      console.log('🎯 Mobile: Getting Agora token for channel:', channelName);
      
      const response = await fetch('http://192.168.86.113:5262/api/agora/token', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${authToken}`,
        },
        body: JSON.stringify({
          channelName: channelName,
          expirationTimeInSeconds: 3600,
        }),
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const data = await response.json();
      console.log('✅ Mobile: Got Agora token:', data.token?.substring(0, 20) + '...');
      
      return data.token;
    } catch (error) {
      console.error('❌ Mobile: Failed to get Agora token:', error);
      throw error;
    }
  }
}

// Export singleton instance
export default new AgoraService();

// Export components for video rendering
export { RtcLocalView, RtcRemoteView, VideoRenderMode };
